src/pybind/mgr/prometheus/module.py

```diff
@@ -1294,15 +1296,17 @@ def get_rbd_stats(self) -> None:
                 namespace_name = s[1]
 
             if pool_name == "*":
-                # collect for all pools
-                osd_map = self.get('osd_map')
-                for pool in osd_map['pools']:
+                # collect for all existing pools
+                for pool in existing_pools:
```

Comment:
Unless strictly required (iterating through a 2D array) in general is good to avoid nested for loops, especially when it's not strictly necessary. Instead, what about:
- Tokenize user-provided pools (no for-loop is even required):
```py
user_provided_pools = list(
  {  # this is a set
    pool.strip()  # to remove whitespaces
      for self.get_module_option('rbd_stats_pools').split(',')  # to tokenize
      if pool  # to remove empty strings
  }
)
```
- Then check if some element is the `*`:
```py
if '*' in user_provided_pools:
```
Now I see that you need the `osd_map` 'cause it's the only map that contains the application_metadata of a pool... Probably that could/should be added to other calls (like `pool_stats` or `df`) instead of exposing that in a heavy-weight API call like the `osd_map` @pereman2 what are your thoughts here?

Reply:
> Unless strictly required (iterating through a 2D array) in general is good to avoid nested for loops, especially when it's not strictly necessary. Instead, what about:
> 
>     * Tokenize user-provided pools (no for-loop is even required):
> 
> 
> ```python
> user_provided_pools = list(
>   {  # this is a set
>     pool.strip()  # to remove whitespaces
>       for self.get_module_option('rbd_stats_pools').split(',')  # to tokenize
>       if pool  # to remove empty strings
>   }
> )
> ```
> 
>     * Then check if some element is the `*`:
> 
> 
> ```python
> if '*' in user_provided_pools:
> ```
> 
> Now I see that you need the `osd_map` 'cause it's the only map that contains the application_metadata of a pool... Probably that could/should be added to other calls (like `pool_stats` or `df`) instead of exposing that in a heavy-weight API call like the `osd_map` @pereman2 what are your thoughts here?

Well for each pool name we anyways have to check if it's present in `osd_map['pools']` or not, so it'll be `O(n)` per `pool_name`. And for improvement intead of all pools now I'm fetching only rbd pools  before the loop.